### PR TITLE
Create a new ledger chunk after public recovery

### PR DESCRIPTION
**Pull request by jumaffre**
First part of #1652 

A new ledger chunk is now created after the public recovery is complete so that there's a clear separation of ledger files between service epochs.

To do so, the primary node (started in `recover` mode) forces a new ledger chunk when truncating the recovered ledger at the end of the public recovery (new `complete` argument to ledger `truncate()`). Because the ledger chunks have to be in sync between all the nodes, the primary passes the `last_recovered_signed_idx` to the backup nodes (this was already the case before this PR) and the backup's ledger is then initialised with this index so that a new chunk can be forced when this index is eventually replicated.

Note that this doesn't make use of the new feature introduced in #3519 as this would require rewriting the historical ledger and thus breaking history.

TODO:
- [x] Cleanup
- [x] Ledger unit tests

### CODE REVIEW

File: src/host/ledger.h [hunk 1]
```diff
@@ -954,17 +970,30 @@ namespace asynchost
       return last_idx;
     }
 
-    void truncate(size_t idx)
+    void truncate(size_t idx, bool complete = false)
```
Comment (wintersteiger) [inline src/host/ledger.h:973]: `complete` could be a little confusing, I'd prefer `complete_chunk`, there's also a `complete_ledger` further down

**Review by wintersteiger — APPROVED** [review]
LGTM. Perhaps test for the new chunk in one of the existing recovery tests?

File: tests/infra/network.py [hunk 2]
```diff
@@ -321,11 +321,18 @@ def _start_all_nodes(
...
                         )
-                        self.wait_for_state(
+                        last_signed_seqno = self.wait_for_state(
                             node,
                             infra.node.State.PART_OF_PUBLIC_NETWORK.value,
                             timeout=args.ledger_recovery_timeout,
                         )
+                        # Verify that a new ledger chunk was created after public recovery
+                        for f in os.listdir(node.remote.ledger_paths()[0]):
+                            start_seqno, end_seqno = Ledger.range_from_filename(f)
```
[the remark this answers is not in the document]
Comment (jumaffre) [reply tests/infra/network.py:331]: I've removed the underscore from the name as we now access it from outside the `Ledger` class. I don't know why the function can return `(False, elements)` so I've removed this path too.

File: src/node/node_state.h [hunk 3]
```diff
@@ -2054,9 +2057,10 @@ namespace ccf
         consensus::LedgerRequestPurpose::Recovery);
     }
 
-    void ledger_truncate(consensus::Index idx)
+    void ledger_truncate(consensus::Index idx, bool complete_ledger = false)
```
[the remark this answers is not in the document]
Comment (jumaffre) [reply src/node/node_state.h:2060]: The enclave has little notion of ledger chunks and I think it's more intuitive to say that the whole ledger is complete at this point, hence why I called it `complete_ledger`.

File: tests/infra/network.py [hunk 4]
```diff
@@ -587,9 +594,7 @@ def stop_all_nodes(self, skip_verification=False, verbose_verification=False):
                 fatal_error_found = True
 
         LOG.info("All nodes stopped")
-
-        if not skip_verification:
-            self.check_ledger_files_identical()
+        self.check_ledger_files_identical()
```
[the remark this answers is not in the document]
Comment (jumaffre) [reply tests/infra/network.py:597]: #3530

**Comment by jumaffre** [comment]
Abandoning this PR as this isn't the way to go (as identified by multiple recoveries tested in test suites). I will open a new PR making use of the ledger entries flag instead.

**Pull request closed without merging**